Adding the LSTM text generation example

keskarnitish opened the pull request:
This adds an example which trains a 2 layered LSTM on a user-provided text file and uses that to generate text. It is partly based on [Andrej Karpathy's blog](karpathy.github.io/2015/05/21/rnn-effectiveness/) and his accompanying file [min-char-rnn.py](https://gist.github.com/karpathy/d4dee566867f8291f086) . When I playing around with [Keras] (keras.io), I noticed it had a similar example and wrote one which uses Lasagne. I realize that Lasagne already has an example for RNNs (The addition problem), but I felt this might be an interesting addition. I have commented it in attempts to make it self-explanatory and have tested this on both CPU and GPU.

(Originally this was a [PR on Lasagne](https://github.com/Lasagne/Lasagne/pull/422#issuecomment-141711874), but @skaee suggested it be requested here.)

skaae commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,202 @@
+    def try_it_out(seed, N=200):
+        '''
+        This function uses the current state of the RNN to generate text. It takes as input the first character, 
+        computes the probability distribution of the next (using the RNN), chooses one at random using the computed probabilities
```
change chooses to samples?

skaae commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,202 @@
+
+                avg_cost += train(x, y)
+            print("Epoch {} average loss = {}".format(it*1.0*PRINT_FREQ/data_size*BATCH_SIZE, avg_cost / PRINT_FREQ))
+        import ipdb; ipdb.set_trace()
```
This should probably removed?

skaae commented:
It looks very nice!

Can you explain the semi redundant dataset and how `try_it_out works`?

If it is not too much work it would be nice to turn it into a notebook where the users can see generated text.

keskarnitish commented:
I reorganized my code so the try_it_out function is a little easier to interpret. I have also added explanation for the semi-redundant batches and try_it_out. I hope that helps. 
Also, I removed the ipdb.set_trace(); I was using it for debugging and forgot to take it out when committing. Lastly, I'm not very experienced at IPython notebooks; I made a rudimentary one [here](https://gist.github.com/keskarnitish/ec4d45ae65fe6b7f1679) which I ran for only a few iterations.

f0k commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,230 @@
+    For the first batch,
+    x (the inputs to the neural network) would correspond to the encoding of 'T','h','e',' ','q'
+    y (the targets of the neural network) would be the encoding of 'u'
+    For the second batch,
```
You mean: `For the first data point,` and `For the second data point,`

keskarnitish replied:
Agreed. I fixed this.

f0k commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,230 @@
+    For the second batch,
+    x (the inputs to the neural network) would correspond to the encoding of 'h','e',' ','q', 'u'
+    y (the targets of the neural network) would be the encoding of 'i'
+    The batches are then stacked (into a three-dimensional tensor) and returned. 
```
And `The data points are then stacked`. What you return is a single batch, i.e., a tensor of multiple data points.

keskarnitish replied:
What I meant to say is that the 2D data points are "stacked" to create a 3D tensor. Maybe, the right word is "aggregated".

f0k replied:
No, stacked is fine! It's just that you're stacking data points into a batch, while the docstring says you're stacking batches. So just replace `batches` by `data points`.

f0k commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,230 @@
+        for i in range(SEQ_LENGTH):
+            x[n,i,char_to_ix[data[p+ptr+i]]] = 1.
+        y[n] = char_to_ix[data[p+ptr+SEQ_LENGTH]]
+    return x, np.array(y,dtype='int32')
```
Not sure if this is a bottleneck, but there'd be several possible optimizations here.
First, you could convert all the characters beforehands:
```
data = np.fromiter((char_to_ix(c) for c in data[p:p+SEQ_LENGTH+batch_size]), dtype=np.int32)
```
Then instead of copying data around, you can just create `x` and `y` directly:
```
x = np.lib.stride_tricks.as_strided(
        data[:-1],
        (batch_size, SEQ_LENGTH),
        (data.strides[0], data.strides[0])))
y = data[1:]
```
You'd save even more if you'd convert the whole text to a numpy array via `np.fromiter` directly after loading it.

f0k replied:
I just see that for `x` you need the one-hot encoded data. Should be doable like this:
```python
# assume data was already converted to an int array on loading
data = data[p:p+SEQ_LENGTH+batch_size]
x = lasagne.utils.one_hot(data[:-1], vocab_size)
x = np.lib.stride_tricks.as_strided(
    x, (batch_size, SEQ_LENGTH, vocab_size),
    (x.strides[0], x.strides[0], x.strides[1]))
y = data[1:]
```
Although it might be more efficient to just keep feeding ints for both `x` and `y` and insert an `EmbeddingLayer` right after the `InputLayer` that is initialized with `np.eye` and made untrainable (`layer.params[layer.W].remove('trainable')`).

skaae commented:
@keskarnitish: Is it correct that the `try_it_out` is inefficient because it needs to process SEQ_LEN*N to generate N outputs? Or maybe I don't understand the implementation :)

Could you run a LSTM for a single time step then sample the output and run another time step?
That would of course require that the cell and hidden values are input at every timestep.

f0k commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,230 @@
+        x,_ = gen_data(0, 1, generation_phrase)
+
+        for i in range(N):
+            ix = np.argmax(probs(x).ravel())
```
Wouldn't it be nicer to sample from the probability distribution, rather than just picking the maximum?

keskarnitish replied:
@skaae, you're right. That is another way of doing it. I was trying to mimic the [Keras](https://github.com/fchollet/keras/blob/master/examples/lstm_text_generation.py) example. 

@f0k, that's how I had it when I first wrote the code. I was using `np.random.choice` instead of `np.argmax`. However, there seemed to be some NumPy issues that were cropping in (especially if the NumPy version was slightly older). I could write my own random choice but ultimately, I felt that `argmax` was interpretable and needed no code.

f0k replied:
Okay, fair enough! Maybe add the second variant in a comment here:
```python
# Pick the character that got assigned the highest probability
ix = np.argmax(probs(x).ravel())
# Alternatively, to sample from the distribution instead:
# ix = np.random.choice(np.arange(vocab_size), p=probs(x).ravel())
```

ebenolson commented:
Thanks for the example!

I haven't really looked at it carefully, but I wonder if there's an off by one error somewhere, because when I ran it, the generated snippets all looked like this
>  The quick brown fox jumpssuch thing as sure, the spirit of the subject of the uppertsions, that man is grateful in the principal that are complete senses it is
desires and pain and all that fashion will that is good in great 

with no space after "jumps", but it seems to be pretty good at generating separated words otherwise?

f0k commented:
Looks promising!

> Is it correct that the try_it_out is inefficient because it needs to process SEQ_LEN*N to generate N outputs?

That's the way I see it, yes.

> Could you run a LSTM for a single time step then sample the output and run another time step?
That would of course require that the cell and hidden values are input at every timestep.

Yes, it would require the network to remember the cell and hidden states between calls, which is not easily possible right now, is it? Would it be possible to extend the recurrent layers' `get_output_for` to accept a keyword argument such as `stateful` or `persist_state` that would tell them to use a shared variable for the hidden and cell states and update that at each call (using the trick I used for the batch normalization layer)? The question is how to reset the state then...

By the way, it should even be possible to integrate all the sampling machinery into a scan, shouldn't it? That'd be even more difficult to provide an API for, though.

ebenolson commented on examples/lstm_text_generation.py:
```diff
@@ -0,0 +1,230 @@
+    p = 0
+    try:
+        for it in xrange(data_size * num_epochs / BATCH_SIZE):
+            sample_ix = try_it_out() # Generate text using the p^th character as the start. 
```
what's this for? `try_it_out` doesn't return anything?

keskarnitish replied:
Oh! You're right. I fixed this.

skaae commented:
> Yes, it would require the network to remember the cell and hidden states between calls, which is not easily possible right now, is it? Would it be possible to extend the recurrent layers' get_output_for to accept a keyword argument such as stateful or persist_state that would tell them to use a shared variable for the hidden and cell states and update that at each call (using the trick I used for the batch normalization layer)? The question is how to reset the state then...

Yes I dont think that is possible right know. Maybe the current solution is a fine work around. I don't think we shoud complicate the RNN layers more to support this :)

> By the way, it should even be possible to integrate all the sampling machinery into a scan, shouldn't it? That'd be even more difficult to provide an API for, though.

Yes, but that would require us to rewrite the RNN code quite a lot. One possible solution is to create a RNN container. The user should then provide the code for running a single step and the container would handle the recurrence. 

This would of course require singlestep LSTM and RNNs layers.

f0k commented:
> Yes, but that would require us to rewrite the RNN code quite a lot. One possible solution is to create a RNN container. The user should then provide the code for running a single step and the container would handle the recurrence.

Sounds interesting. What would that look like from a user's point of view? I.e., what would become of the following code?
```python
    layer = lasagne.layers.InputLayer(shape=(N_BATCH, MAX_LENGTH, 2))
    layer = lasagne.layers.RecurrentLayer(layer, N_HIDDEN)
    layer = lasagne.layers.SliceLayer(layer, -1, 1)
    layer = lasagne.layers.DenseLayer(layer, num_units=1, nonlinearity=lasagne.nonlinearities.tanh)
```
(I'm aware that this is not the best place to discuss this, but I'd like to see if it's worth creating a Lasagne issue for that.)

skaae commented:
Something like

```Python
    layer = lasagne.layers.InputLayer(shape=(N_BATCH, MAX_LENGTH, 2))

   # single time step
    rnn_layer = lasagne.layers.InputLayer((N_batch, 2))
    rnn_layer = lasagne.layers.LSTMSingleStepLayer(rnn_layer, 100)
    #rnn_layer = lasagne.layers.SampleLayer(rnn_layer)  - could be used for the char-rnn example

    layer = lasagne.layers.RNNContainer(layer, single_time_step=rnn_layer, feedback=False)
    layer = lasagne.layers.SliceLayer(layer, -1, 1)
    layer = lasagne.layers.DenseLayer(layer, num_units=1, nonlinearity=lasagne.nonlinearities.tanh)
```

I addded a feedback flag which should tell the model if you want the previous prediction used as input in the next time step or not.

ebenolson commented:
Looking good!

I think it's ready to merge if you could just address this last comment:

> /edit: One thing I forgot: I'm not so sure about including the text file in the repository, other examples just download their datasets on-demand. If we decide that it's small enough to be included, it should probably get a more descriptive name than input.txt, though, or move it into a subdirectory along with the code. Right now it's not clear which example the text file belongs to.

Either way (download when run or make a subdirectory for this example) is fine by me.

keskarnitish commented:
How about using `urllib2` to read the Nietzsche data set (or something from Project Gutenberg if we want to ensure persistence of the link) with the option of using your own file in comments?

Something like:
<pre>
<code>
import urllib2
in_text = urllib2.urlopen('https://s3.amazonaws.com/text-datasets/nietzsche.txt').read()
#You can also use your own file
#The file must be a simple text file.
#Simply edit the file name below and uncomment the line.  
#in_text = open('your_file.txt', 'r').read()
in_text = in_text.decode("utf-8-sig").encode("utf-8")
</code>
</pre>

Instead of what I have at the moment, which is
<pre>
<code>
fname = 'input.txt' 
in_text = open(fname, 'r').read() # should be simple plain text file
in_text = in_text.decode("utf-8-sig").encode("utf-8")
</code>
</pre>

ebenolson commented:
That sounds great.

ebenolson commented:
Looks good, would you mind removing the `input.txt` file and then squashing into one commit?

f0k commented:
> Looks good, would you mind removing the `input.txt` file and then squashing into one commit?

Just to make it easier, this would be:
```
git reset --soft 589ac63  # place HEAD at first commit of this PR, keep changes staged
git rm input.txt  # additionally remove input.txt
git commit --amend --no-edit  # add all changes to the first commit of this PR
git push --force  # overwrite PR with the new single-commit history
```

ebenolson commented:
Merging, thanks again!
